File: helpers/macos/Sources/SidecarKit/Protocol/SidecarError.swift
import Foundation

/// The v1 error codes (`contracts/sidecar/protocol-v1.md`). The JVM client coerces an unknown code to
/// `unknown`, so a helper may emit others; these are the ones the helper produces.
public enum SidecarErrorCode: String {
    /// The helper does not implement `method`.
    case unknownMethod = "unknown_method"
    /// `params` were missing or malformed.
    case invalidParams = "invalid_params"
    /// Handshake rejected (bad/absent token). Connection-level (`failure` with no `id`).
    case unauthenticated = "unauthenticated"
    /// The capability exists but isn't available now (permission denied, engine busy).
    case unavailable = "unavailable"
    /// The helper failed internally.
    case `internal` = "internal"
    /// A protocol violation (unframeable/oversize/unexpected frame).
    case protocolViolation = "protocol"
    /// Tolerant-decode fallback for an unrecognized code (the helper never *emits* this; it mirrors the
    /// JVM client's `coerceInputValues` so decoding a future code can't throw).
    case unknown = "unknown"
}

/// A wire error object: `{ "code", "message", "details"? }`. `message` is a **non-PII** human-readable
/// summary; `details` is opaque (and, like every payload, must never be logged — ADR-0009).
public struct SidecarError: Equatable {
    public let code: SidecarErrorCode
    public let message: String
    public let details: JSONValue?

    public init(_ code: SidecarErrorCode, _ message: String, details: JSONValue? = nil) {
        self.code = code
        self.message = message
        self.details = details
    }
}

File: core/sidecar/src/main/kotlin/com/circuitstitch/deferno/core/sidecar/SidecarError.kt
package com.circuitstitch.deferno.core.sidecar

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * The Sidecar protocol's error model (ADR-0024). Carried by [SidecarFrame.Failure] — either correlated
 * to a request/stream (`id` set) or connection-level (`id` null). Shaped to mirror core/network's
 * `ApiErrorBody` (a stable `code` + human-readable `message`) so the two error surfaces feel alike.
 *
 * **Privacy (ADR-0009):** [message] is a **non-PII** human-readable summary (like `SpeechError`'s
 * non-PII reasons) — a Helper must never put Transcript text or other private payload in it. [details]
 * is opaque and **redacted from [toString]** so an accidental log can't leak it.
 */
@Serializable
data class SidecarError(
    val code: SidecarErrorCode = SidecarErrorCode.UNKNOWN,
    val message: String = "",
    val details: JsonElement? = null,
) {
    override fun toString(): String = "SidecarError(code=$code, message=$message, details=${redact(details)})"
}

/**
 * The closed set of Sidecar error codes. Decoding is **tolerant** (ADR-0005): a code a newer Helper
 * sends that this client doesn't know coerces to [UNKNOWN] (via `coerceInputValues` in [SidecarJson]),
 * so a forward-compatible Helper never breaks the client.
 */
@Serializable
enum class SidecarErrorCode {
    /** The requested method is not implemented by the bound Helper. */
    @SerialName("unknown_method") UNKNOWN_METHOD,

    /** The request's params were missing or malformed. */
    @SerialName("invalid_params") INVALID_PARAMS,

    /** The peer-auth handshake failed — bad/absent token (the client half, ADR-0024). */
    @SerialName("unauthenticated") UNAUTHENTICATED,

    /** The capability exists but is not available right now (e.g. permission denied, engine busy). */
    @SerialName("unavailable") UNAVAILABLE,

    /** The Helper hit an internal error fulfilling the request. */
    @SerialName("internal") INTERNAL,

    /** A protocol violation — unframeable bytes, an unexpected frame, a correlation-id clash. */
    @SerialName("protocol") PROTOCOL,

    /** Tolerant fallback for a code this client does not recognize (forward-compat). */
    @SerialName("unknown") UNKNOWN,
}
